Add beatmap drain time check

NiceAesth opened the pull request:
Related: #12091 

This PR implements the minimum drain time check.

I intend to implement more checks, but I decided to start with a simpler one in order to see how to properly do so.

![image](https://github.com/ppy/osu/assets/8437201/97e118d5-b7e2-4a64-9e44-0961c2ced744)

NiceAesth committed:
Add beatmap minimum length checks

NiceAesth committed:
Add tests

NiceAesth commented on osu.Game.Tests/Editing/Checks/CheckDrainTimeTest.cs:
```diff
@@ -0,0 +1,50 @@
+            {
+                HitObjects =
+                {
+                    new HitCircle()
```
I am not entirely sure if this is the correct way to approach this.

bdach replied:
I'm not sure how to respond this without knowing what you're specifically unsure about.

bdach commented on osu.Game/Rulesets/Edit/Checks/CheckDrainTime.cs:
```diff
@@ -0,0 +1,38 @@
+
+        public IEnumerable<Issue> Run(BeatmapVerifierContext context)
+        {
+            double drainTime = context.Beatmap.CalculatePlayableLength();
```
This is not what drain time is. Drain time is, as the name suggests, the amount of time during gameplay wherein HP drain is active. HP drain is not active during breaks. `CalculatePlayableLength()` does not account for breaks in any way, shape, or form.

Compare implementation in [Mapset](https://github.com/Naxesss/MapsetChecks/blob/8efbf76d17aa495732b2de0b75e33132169cecd4/Checks/AllModes/Compose/CheckDrainTime.cs#L55) [Verifier](https://github.com/Naxesss/MapsetParser/blob/8f14354073ed9f2c4723bbd96ac8602cdc6693b6/objects/Beatmap.cs#L606).

NiceAesth replied:
You are right about this, apologies for the oversight. I will get back to adressing the feeeback soon, but I assume that drain time is not currently calculated (or at least I cannot find the relevant code for now). Should the calculations also be added to beatmap extensions in case they become reused for future UI?

bdach commented on osu.Game.Tests/Editing/Checks/CheckDrainTimeTest.cs:
```diff
@@ -0,0 +1,50 @@
+        [Test]
+        public void TestDrainTimeShort()
+        {
+            setShortDrainTimeBeatmap();
+            var content = new BeatmapVerifierContext(beatmap, new TestWorkingBeatmap(beatmap));
```
So `setShortDrainTimeBeatmap()` mutates a field of the test, which then is passed to `BeatmapVerifierContext` via the field. This muddles cause and effect when reading this code.

Please change `void setShortDrainTimeBeatmap()` to `IBeatmap getShortDrainTimeBeatmap()` and pass that to `BeatmapVerifierContext` explicitly.

NiceAesth replied:
This was what I initially had, but I changed it after comparing to https://github.com/cdwcgt/osu/blob/master/osu.Game.Tests/Editing/Checks/CheckPreviewTimeTest.cs

I will change my PR to use the mentioned method(s) instead.

bdach commented on osu.Game.Tests/Editing/Checks/CheckDrainTimeTest.cs:
There should be positive tests cases in this file too (as in, it should be tested that the check is not added on beatmaps with drain time above 30 seconds).

After the part wherein the current implementation is wrong for not accounting for breaks, the "accounting for breaks" part should have test coverage, too.

NiceAesth committed:
Account for break time

NiceAesth committed:
Apply feedback to tests

NiceAesth commented:
I have updated the PR to reflect the changes mentioned in the given feedback

bdach committed:
Rewrite tests to be less aggressively DRY

Four of six helper methods defined in the test were used exactly once;
the remaining two were used two times. Splitting helpers there is just
too much.

bdach committed:
Actually cover cases with breaks in tests

bdach committed:
Improve negative test case without breaks too

bdach committed:
Merge branch 'master' into verify-drain-time

bdach commented:
I basically had to rewrite the test cases because they were so aggressively DRY that it was starting to get difficult to actually follow what the tests were testing, which should never happen given how simple the tests actually are. In tests, when in doubt, [DAMP over DRY](https://enterprisecraftsmanship.com/posts/dry-damp-unit-tests/).

Other than that, seems to work. The check may behave a little wonky on legacy beatmaps with breaks because lazer editor can't manipulate or generate breaks yet, but that's tracked [elsewhere](https://github.com/ppy/osu/issues/14330).
